ddl/partition.go

```diff
@@ -1184,14 +1185,30 @@ func buildRangePartitionDefinitions(ctx sessionctx.Context, defs []*ast.Partitio
 
 		buf := new(bytes.Buffer)
 		// Range columns partitions support multi-column partitions.
-		for _, expr := range clause.Exprs {
+		for i, expr := range clause.Exprs {
 			expr.Accept(exprChecker)
 			if exprChecker.err != nil {
 				return nil, exprChecker.err
 			}
-			expr.Format(buf)
-			piDef.LessThan = append(piDef.LessThan, buf.String())
-			buf.Reset()
+			if len(partValStrings) > i {
+				partVal := partValStrings[i]
+				switch colTypes[i].EvalType() {
+				case types.ETInt:
+					// no wrapping
+				case types.ETDatetime, types.ETString:
+					if _, ok := clause.Exprs[i].(*ast.MaxValueExpr); !ok {
+						// Don't wrap MAXVALUE
+						partVal = driver.WrapInSingleQuotes(partVal)
+					}
+				default:
+					return nil, dbterror.ErrWrongTypeColumnValue.GenWithStackByArgs()
+				}
+				piDef.LessThan = append(piDef.LessThan, partVal)
+			} else {
```

Comment:
Why add this, instead of keeping the original logic?

Reply:
Try the following test:
```sql
 create table t(a time) partition by range columns (a) (partition p1 values less than ('2020'), partition p2 values less than ('20:20:10'));
```

MySQL will execute successfully but TiDB will be failed.

Reply:
Good find!
Looking at https://dev.mysql.com/doc/refman/5.7/en/partitioning-columns.html, `time` is not documented to be supported. But since it does work in 6.2 and I have broken the single-column RANGE COLUMNS implementation for this case I will look into it and fix it.

The reason for this is that the code earlier (`checkAndGetColumnsTypeAndValuesMatch`) already transforms and evaluates the expression, so we can get the value string from there directly.